#### 1. Import pandas library

In [71]:
import pandas as pd
import getpass 

#### 2. Import pymysql and sqlalchemy as you have learnt in the lesson of importing/exporting data 


In [72]:
import pymysql
from sqlalchemy import create_engine 

#### 3. Create a mysql engine to set the connection to the server. Check the connection details in [this link](https://relational.fit.cvut.cz/search?tableCount%5B%5D=0-10&tableCount%5B%5D=10-30&dataType%5B%5D=Numeric&databaseSize%5B%5D=KB&databaseSize%5B%5D=MB)

In [73]:
#engine = db.create_engine('dialect+driver://user:pass@host:port/db')

password = getpass.getpass("Insert your mysql root password: ")
engine = create_engine('mysql+pymysql://guest:{}@relational.fit.cvut.cz/stats'.format(password))
print("Connected to server!")

Insert your mysql root password: ········
Connected to server!


#### 4. Import the users table 

In [74]:
users = pd.read_sql_query("SELECT * FROM users",engine)

In [75]:
users.head()

Id  Reputation        CreationDate   DisplayName      LastAccessDate  \
0  -1           1 2010-07-19 06:55:26     Community 2010-07-19 06:55:26   
1   2         101 2010-07-19 14:01:36  Geoff Dalgas 2013-11-12 22:07:23   
2   3         101 2010-07-19 15:34:50  Jarrod Dixon 2014-08-08 06:42:58   
3   4         101 2010-07-19 19:03:27        Emmett 2014-01-02 09:31:02   
4   5        6792 2010-07-19 19:03:57         Shane 2014-08-13 00:23:47   

                       WebsiteUrl            Location  \
0  http://meta.stackexchange.com/  on the server farm   
1        http://stackoverflow.com       Corvallis, OR   
2        http://stackoverflow.com        New York, NY   
3    http://minesweeperonline.com   San Francisco, CA   
4         http://www.statalgo.com        New York, NY   

                                             AboutMe  Views  UpVotes  \
0  <p>Hi, I'm not really a person.</p>\n\n<p>I'm ...      0     5007   
1  <p>Developer on the StackOverflow team.  Find ...     25        3   
2  <p><a href="http://blog.stackoverflow.com/2009...     22       19   
3  <p>currently at a startup in SF</p>\n\n<p>form...     11        0   
4  <p>Quantitative researcher focusing on statist...   1145      662   

   DownVotes  AccountId   Age                     ProfileImageUrl  
0       1920         -1   NaN                                None  
1          0          2  37.0                                None  
2          0          3  35.0                                None  
3          0       1998  28.0  http://i.stack.imgur.com/d1oHX.jpg  
4          5      54503  35.0                                None

#### 5. Rename Id column to userId

In [84]:
users_2 = users.rename(columns= {'Id':'userId'})

#### 6. Import the posts table. 

In [85]:
posts = pd.read_sql_query("SELECT * FROM posts",engine)

In [78]:
posts.head()

Id  PostTypeId  AcceptedAnswerId         CreaionDate  Score  ViewCount  \
0   1           1              15.0 2010-07-19 19:12:12     23     1278.0   
1   2           1              59.0 2010-07-19 19:12:57     22     8198.0   
2   3           1               5.0 2010-07-19 19:13:28     54     3613.0   
3   4           1             135.0 2010-07-19 19:13:31     13     5224.0   
4   5           2               NaN 2010-07-19 19:14:43     81        NaN   

                                                Body  OwnerUserId  \
0  <p>How should I elicit prior distributions fro...          8.0   
1  <p>In many different statistical methods there...         24.0   
2  <p>What are some valuable Statistical Analysis...         18.0   
3  <p>I have two groups of data.  Each with a dif...         23.0   
4  <p>The R-project</p>\n\n<p><a href="http://www...         23.0   

      LasActivityDate                                              Title  ...  \
0 2010-09-15 21:08:26                      Eliciting priors from experts  ...   
1 2012-11-12 09:21:54                                 What is normality?  ...   
2 2013-05-27 14:48:36  What are some valuable Statistical Analysis op...  ...   
3 2010-09-08 03:00:19  Assessing the significance of differences in d...  ...   
4 2010-07-19 19:21:15                                               None  ...   

  AnswerCount  CommentCount  FavoriteCount  LastEditorUserId  \
0         5.0             1           14.0               NaN   
1         7.0             1            8.0              88.0   
2        19.0             4           36.0             183.0   
3         5.0             2            2.0               NaN   
4         NaN             3            NaN              23.0   

         LastEditDate  CommunityOwnedDate ParentId  ClosedDate  \
0                 NaT                 NaT      NaN         NaT   
1 2010-08-07 17:56:44                 NaT      NaN         NaT   
2 2011-02-12 05:50:03 2010-07-19 19:13:28      NaN         NaT   
3                 NaT                 NaT      NaN         NaT   
4 2010-07-19 19:21:15 2010-07-19 19:14:43      3.0         NaT   

  OwnerDisplayName LastEditorDisplayName  
0             None                  None  
1             None                  None  
2             None                  None  
3             None                  None  
4             None                  None  

[5 rows x 21 columns]

#### 7. Rename Id column to postId and OwnerUserId to userId

In [86]:
posts_2 = posts.rename(columns={'Id':'postId', 'OwnerUserId':'userId'})

#### 8. Define new dataframes for users and posts with the following selected columns:
    **users columns**: userId, Reputation,Views,UpVotes,DownVotes
    **posts columns**: postId, Score,userID,ViewCount,CommentCount

In [87]:
users_3 = users_2[['userId', 'Reputation', 'Views', 'UpVotes', 'DownVotes']]
posts_3 = posts_2[['postId', 'Score', 'userId', 'ViewCount', 'CommentCount']]

In [88]:
users_3.head()

userId  Reputation  Views  UpVotes  DownVotes
0      -1           1      0     5007       1920
1       2         101     25        3          0
2       3         101     22       19          0
3       4         101     11        0          0
4       5        6792   1145      662          5

In [89]:
posts_3.head()
posts_3[]

SyntaxError: invalid syntax (<ipython-input-89-c6053d3393fa>, line 2)

#### 8. Merge both dataframes, users and posts. 
You will need to make a [merge](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) of posts and users dataframes.

In [90]:
merge_df = pd.merge(users_3, posts_3, on = 'userId')

In [91]:
merge_df.head()

userId  Reputation  Views  UpVotes  DownVotes  postId  Score  ViewCount  \
0      -1           1      0     5007       1920    2175      0        NaN   
1      -1           1      0     5007       1920    8576      0        NaN   
2      -1           1      0     5007       1920    8578      0        NaN   
3      -1           1      0     5007       1920    8981      0        NaN   
4      -1           1      0     5007       1920    8982      0        NaN   

   CommentCount  
0             0  
1             0  
2             0  
3             0  
4             0

#### 9. How many missing values do you have in your merged dataframe? On which columns?

In [92]:
null_cols = merge_df.isnull().sum()
null_cols[null_cols > 0]

ViewCount    48396
dtype: int64

#### 10. You will need to make something with missing values.  Will you clean or filling them? Explain. 
**Remember** to check the results of your code before passing to the next step

In [93]:
print('% de valores nulos= ',str(merge_df['ViewCount'].isnull().sum()/merge_df.shape[0]*100)[:4],'%')

% de valores nulos=  53.4 %


Voy a rellenar los valores nulos con cero porque 'sólo' el 53.4% son nulos y creo que es relevante
ekl kpi para posibles análisis

In [94]:
merge_df[['ViewCount']] = merge_df[['ViewCount']].fillna(0)

In [95]:
merge_df.head()

userId  Reputation  Views  UpVotes  DownVotes  postId  Score  ViewCount  \
0      -1           1      0     5007       1920    2175      0        0.0   
1      -1           1      0     5007       1920    8576      0        0.0   
2      -1           1      0     5007       1920    8578      0        0.0   
3      -1           1      0     5007       1920    8981      0        0.0   
4      -1           1      0     5007       1920    8982      0        0.0   

   CommentCount  
0             0  
1             0  
2             0  
3             0  
4             0

#### 11. Adjust the data types in order to avoid future issues. Which ones should be changed? 

In [96]:
merge_df.dtypes

userId            int64
Reputation        int64
Views             int64
UpVotes           int64
DownVotes         int64
postId            int64
Score             int64
ViewCount       float64
CommentCount      int64
dtype: object

In [97]:
merge_df['userId'] = merge_df['userId'].astype('str')
merge_df['postId'] = merge_df['postId'].astype('str')
merge_df['ViewCount'] = merge_df['postId'].astype('int64')

In [98]:
merge_df.dtypes

userId          object
Reputation       int64
Views            int64
UpVotes          int64
DownVotes        int64
postId          object
Score            int64
ViewCount        int64
CommentCount     int64
dtype: object

#### Bonus: Identify extreme values in your merged dataframe as you have learned in class, create a dataframe called outliers with the same columns as our data set and calculate the bounds. The values of the outliers dataframe will be the values of the merged_df that fall outside that bounds. You will need to save your outliers dataframe to a csv file on your-code folder.

In [99]:
stats = merge_df.describe().transpose()
stats['IQR'] = stats['75%'] - stats['25%']
stats

count          mean           std   min       25%      50%  \
Reputation    90584.0   6282.395412  15102.268670   1.0     60.00    396.0   
Views         90584.0   1034.245176   2880.074012   0.0      5.00     45.0   
UpVotes       90584.0    734.315718   2050.869327   0.0      1.00     22.0   
DownVotes     90584.0     33.273249    134.936435   0.0      0.00      0.0   
Score         90584.0      2.780767      4.948922 -19.0      1.00      2.0   
ViewCount     90584.0  56539.080522  33840.307529   1.0  26051.75  57225.5   
CommentCount  90584.0      1.894650      2.638704   0.0      0.00      1.0   

                   75%       max       IQR  
Reputation     4460.00   87393.0   4400.00  
Views           514.25   20932.0    509.25  
UpVotes         283.00   11442.0    282.00  
DownVotes         8.00    1920.0      8.00  
Score             3.00     192.0      2.00  
ViewCount     86145.25  115378.0  60093.50  
CommentCount      3.00      45.0      3.00

In [102]:
outliers = pd.DataFrame(columns=merge_df.columns)

for col in stats.index:
    iqr = stats.at[col,'IQR']
    cutoff = iqr * 1.5
    lower = stats.at[col,'25%'] - cutoff
    upper = stats.at[col,'75%'] + cutoff
    results = merge_df[(merge_df[col] < lower) | 
                   (merge_df[col] > upper)].copy()
    results['Outlier'] = col
    outliers = outliers.append(results)

/Users/mariaadradosreig/Library/Python/3.7/lib/python/site-packages/pandas/core/frame.py:7138: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort,


In [103]:
outliers.head()

CommentCount DownVotes     Outlier Reputation Score UpVotes ViewCount  \
1155            0       126  Reputation      14082    25    4235        74   
1156            0       126  Reputation      14082     5    4235        94   
1157            1       126  Reputation      14082     7    4235        99   
1158            3       126  Reputation      14082     6    4235       119   
1159            0       126  Reputation      14082     7    4235       140   

     Views postId userId  
1155  3320     74     88  
1156  3320     94     88  
1157  3320     99     88  
1158  3320    119     88  
1159  3320    140     88

In [105]:
merge_df.to_csv(r'./outliers.csv')
